### PR TITLE
Add cyclic module import detection support

### LINKED ISSUE #4878: Detect package import <-> package definition cyclic loops
**Description:**
If a source code contains something like the following, the program just hangs. No warning, no errors, doesn't exit. This makes it hard to debug since you have no clue as to what's going on.
```ballerina
package ballerina.net.http;

import ballerina.net.http;
```

### LINKED ISSUE #15031: Parser crashing with EmptyStackException
**Description:**
BallerinaParser is crashing while trying to parse an import from the same module. The attached zip contains a minimal project to reproduce and the `ballerina-internal.log` in case it cannot be reproduced.

```
ERROR {org.ballerinalang.launcher.Main} - null 
java.util.EmptyStackException
	at java.util.Stack.peek(Stack.java:102)
	at java.util.Stack.pop(Stack.java:84)
	at org.wso2.ballerinalang.compiler.parser.BLangPackageBuilder.addExpressionStmt(BLangPackageBuilder.java:2590)
	at org.wso2.ballerinalang.compiler.parser.BLangParserListener.exitExpressionStmt(BLangParserListener.java:1827)
	at org.wso2.ballerinalang.compiler.parser.antlr4.BallerinaParser$ExpressionStmtContext.exitRule(BallerinaParser.java:9308)
	at org.antlr.v4.runtime.Parser.triggerExitRuleEvent(Parser.java:430)
	at org.antlr.v4.runtime.Parser.exitRule(Parser.java:646)
...
```

**Steps to reproduce:**
Attached minimal project to reproduce, just run `ballerina run main.bal`.

**Affected Versions:** 0.990.3

**OS, DB, other environment details and versions:** Ubuntu 18.04
[bal_crash_proj.zip](https://github.com/ballerina-platform/ballerina-lang/files/3076770/bal_crash_proj.zip)

### LINKED ISSUE #15035: Identify cyclic dependencies in imports
**Description:**

Currently, there is no way to identify cyclic dependencies in imports. So when we have a project like below, the compiler crashes.

```
├── abc (imports def)
├── def (imports abc)
└── Ballerina.toml
```

### PR DESCRIPTION
**Pull request by Shan1024**
## Purpose
This PR adds cyclic module imports detection support to the compiler.

Fixes #4878
Fixes #15031
Fixes #15035

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Required Balo version update
- [ ] Updated Change Log
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by MaryamZi** [comment]
IIRC #4878 was maintained to detect and fix issues with cyclic imports too.

e.g.,
```
├── abc (imports def)
├── def (imports abc)
└── Ballerina.toml
```

We wouldn't be fixing that particular scenario with these changes right? Shall we keep 4878 open or open a separate issue if not?

**Comment by Shan1024** [comment]
@MaryamZi - I think it is better if we handle these issues separately. So I created a #15035 to track cyclic imports issue.

**Comment by Shan1024** [comment]
@MaryamZi - This PR has been updated to identify cyclic module imports. This is done in this PR itself is because self import can be considered as a cyclic import as well.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 1]
```diff
@@ -321,6 +330,47 @@ public void visit(BLangImportPackage importPkgNode) {
             return;
         }
 
+        // Detect cyclic module dependencies. This will not detect cycles which starts with the entry package because
+        // entry package has a version. So we check import cycles which starts with the entry package in next step.
+        if (importedPackages.contains(pkgId)) {
+            int index = importedPackages.indexOf(pkgId);
+            // Generate the import cycle.
+            StringBuilder stringBuilder = new StringBuilder();
```
Comment (pubudu91) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: The code for import cycle generation is duplicated ne? Shall we extract that out as a separate method?

Comment (Shan1024) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: This has been addressed with https://github.com/ballerina-platform/ballerina-lang/pull/15230.

**Pull request merged**